### PR TITLE
Support `None` as a default value in `rule`, `repository_rule` and `tag_class` attributes

### LINKED ISSUE #20744: tag_classes don't support optional attrs
### Description of the feature request:

In [migrating rules_rust's third-party resolver support](https://github.com/bazelbuild/rules_rust/pull/2392) to bzlmod we've run into an issue due to `tag_class` arguments being more typed than WORKSPACE functions were.

We want to support a structure along the lines of (WORKSPACE):

```starlark
crates_repostory(
    name = "crates",
    cargo_lockfile = "//third-party:Cargo.lock",
    manifests = ["//third-party:Cargo.toml"],
    default_generate_build_scripts = True,
    annotations = {
        "some-crate": [
            crate.annotation(generate_build_scripts = False),
        ],
    }
)
```

where `crate.annotation` is just a function with optional arguments which serializes its args as JSON (https://github.com/bazelbuild/rules_rust/blob/40232a07f11376c737b35bbfac89773880d09048/crate_universe/private/crate.bzl#L84-L212)

This allowed setting a default boolean to be used for all crates, and overriding it for specific crates.

In bzlmod, we're using a `tag_class` for these annotations, along the lines of:
```starlark
crate.from_cargo(
    name = "crates",
    cargo_lockfile = "//third-party:Cargo.lock",
    manifests = ["//third-party:Cargo.toml"],
    default_generate_build_scripts = True,
)

crate.annotation(
    crate = "some-crate",
    generate_build_scripts = False,
)
```

However, `tag_class`es take fully typed arguments (which is generally good) and don't allow for `None` values, so there doesn't appear to be a way for us to say "`generate_build_scripts` is an optional argument which may be `None` in which case we should fall back to the repo's default value". Because we need to set a non-`None` default for `crate.annotation`'s `generate_build_scripts` argument which means it will always have a value and never fall back to the repo's default.

As a workaround, we can use a string argument here (and perhaps copy rules_go's convention of using strings which are "on"/"off"/"auto" and default to "auto", rather than using a None-able boolean), but this feels kind of messy and hacky.

### Which category does this issue belong to?

External Dependency, Rules API

### What underlying problem are you trying to solve with this feature?

Allowing a bzlmod repo to set a default value for all of the third-party libraries which it generates rules, which may optionally be overridden in for some subset of those libraries.

### Which operating system are you running Bazel on?

_No response_

### What is the output of `bazel info release`?

_No response_

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

_No response_

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

_No response_

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

### PR DESCRIPTION
**Pull request by Silic0nS0ldier**
This PR changes `attr.*(..., default = ..., ...)` so that `default` accepts `None`, and uses that as the default instead of an implicit or explicit type that matches.

e.g.
```starlark
def _impl(ctx):
    print(ctx.attr.bar)
    # Prints: None
    return []

foo = rule(
    implementation = _impl,
    attrs = {
        "bar": attr.string(default = None),
    },
)
```

Note that `rule`s, `repository_rule`s and `tag_class`es (with #23547) will use the default value (implicit or explicit) if `None` is given for an attribute. That is, `None` can only show up if `default = None`.

-----

This could be taken further by adding something like `allow_none` so that an attribute can accept `None` while still defaulting to something else, however this would have one significant drawback (which I'll phrase 3 different ways to help capture the scope).
1. macros (and more generally, functions) will become a source for errors, as default values (`foo = None`) and argument collectors (`**kwargs`) would have a subtle behaviour change when interacting with `rule`s, `repository_rule`s and `tag_class`es.
2. The existence of fields in a `dict` will matter more.
3. Users can no longer rely on `foo = None` to explicitly mean "use the default". This also means it is harder to document _why_ a default may need to be used in a given place.

TL;DR
- `default = None` is scoped, there is little of consequence and rule authors can use it in a way that is not a breaking change (e.g. opt-in change of default value, like with `crate.from_cargo` and `crate.annotation` in Rules Rust).
- `allow_none = True, default = "something-else"` means users need to consider `None` _and_ attribute presence in argument spreads. The change can be released in a non-breaking manner, but not as easily and with a greater risk a mistake being made.

-----

Needs #23547
Resolves #20744

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/packages/Attribute.java [hunk 1]
```diff
@@ -665,6 +666,23 @@ public Builder<TYPE> value(LateBoundDefault<?, ? extends TYPE> defaultValue) {
       return this;
     }
 
+    /**
+     * Sets the attribute default value. The type of the default value must match the type
```
Comment (Silic0nS0ldier) [inline src/main/java/com/google/devtools/build/lib/packages/Attribute.java:670]: TODO Update doc to reflect implementation

File: src/main/java/com/google/devtools/build/lib/packages/AttributeMap.java [hunk 2]
```diff
@@ -57,6 +57,8 @@ default String describeRule() {
    *
    * <p>If the rule doesn't have this attribute with the specified type, throws an
    * {@link IllegalArgumentException}.
+   * 
+   * Note that null may be the default value, check if default set
```
Comment (Silic0nS0ldier) [inline src/main/java/com/google/devtools/build/lib/packages/AttributeMap.java:61]: TODO Rework doc

File: src/main/java/com/google/devtools/build/lib/packages/AttributeMap.java [hunk 3]
```diff
@@ -65,6 +67,8 @@ default String describeRule() {
    * Returns the value of the named rule attribute if it exists, otherwise the given default value.
    * This may be null (for example, for an attribute with no default value that isn't explicitly set
    * in the rule - see {@link Type#getDefaultValue}).
+   *
+   * Note that null may be the default value, check if default set
```
Comment (Silic0nS0ldier) [inline src/main/java/com/google/devtools/build/lib/packages/AttributeMap.java:71]: TODO Rework doc (and maybe point to `get` to cut down on duplication)

File: src/main/java/com/google/devtools/build/lib/starlarkbuildapi/StarlarkAttrModuleApi.java [hunk 4]
```diff
@@ -128,7 +128,8 @@ public interface StarlarkAttrModuleApi extends StarlarkValue {
   String DEFAULT_ARG = "default";
   // A trailing space is required because it's often prepended to other sentences
   String DEFAULT_DOC =
-      "A default value to use if no value for this attribute is given when instantiating the rule.";
+      "A default value to use if no value for this attribute is given when instantiating the rule. "
+          + "<code>None</code> is allowed, and can be used to identify an unset attribute.";
```
Comment (Silic0nS0ldier) [inline src/main/java/com/google/devtools/build/lib/starlarkbuildapi/StarlarkAttrModuleApi.java:132]: TODO Needs some more work. And would likely benefit from some formatting so `None` behaviour is easier to see at a glance.

File: src/test/java/com/google/devtools/build/lib/packages/RuleClassTest.java [hunk 5]
```diff
@@ -1297,4 +1297,43 @@ public void testPackageMetadataAlternateName() throws Exception {
     assertThat(noopRule.getAttr(RuleClass.APPLICABLE_METADATA_ATTR, LABEL_LIST))
         .isEqualTo(Lists.newArrayList(Label.parseCanonical("//testpackage:info")));
   }
+
+  @Test
+  public void testAttr_noneAsDefault() throws Exception {
```
Comment (Silic0nS0ldier) [inline src/test/java/com/google/devtools/build/lib/packages/RuleClassTest.java:1302]: TODO Clean this up

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/TypeCheckedTag.java
Comment (Silic0nS0ldier) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/TypeCheckedTag.java:1]: TODO Remove changes in this file (and those which are related) since #23547 is not going to merge.

File: src/main/java/com/google/devtools/build/lib/packages/AttributeMap.java (hunk 2 above)
Comment (Silic0nS0ldier) [reply src/main/java/com/google/devtools/build/lib/packages/AttributeMap.java:61]: Actually, not completely. It refers to `Type#getDefaultValue` however this is handled by the `Attribute` definition itself (has `valueSet ? value : type.getDefaultValue(),`).
